lua/nvim-tree/watcher.lua

```diff
@@ -16,87 +22,129 @@ local FS_EVENT_FLAGS = {
-    )
-    return nil
+  self._fs_event, _, name = uv.new_fs_event()
+  if not self._fs_event then
+    self._fs_event = nil
+    utils.notify.warn(string.format("Could not initialize an fs_event watcher for path %s : %s", self._path, name))
+    return false
   end
 
-  local event_cb = vim.schedule_wrap(function(err, filename, events)
+  local event_cb = vim.schedule_wrap(function(err, filename)
     if err then
-      log.line("watcher", "event_cb for %s fail : %s", self._opts.absolute_path, err)
+      log.line("watcher", "event_cb for %s fail : %s", self._path, err)
     else
-      log.line("watcher", "event_cb '%s' '%s' %s", self._opts.absolute_path, filename, vim.inspect(events))
-      self._opts.on_event(self._opts)
+      log.line("watcher", "event_cb '%s' '%s'", self._path, filename)
+      for _, listener in ipairs(self._listeners) do
+        listener()
+      end
     end
   end)
 
-  rc, _, name = self._e:start(self._opts.absolute_path, FS_EVENT_FLAGS, event_cb)
+  rc, _, name = self._fs_event:start(self._path, FS_EVENT_FLAGS, event_cb)
   if rc ~= 0 then
-    utils.notify.warn(
-      string.format("Could not start the fs_event watcher for path %s : %s", self._opts.absolute_path, name)
-    )
-    return nil
+    utils.notify.warn(string.format("Could not start the fs_event watcher for path %s : %s", self._path, name))
+    return false
   end
 
-  return self
+  return true
 end
 
-function Watcher:destroy()
-  log.line("watcher", "Watcher:destroy '%s'", self._opts.absolute_path)
-  if self._e then
-    local rc, _, name = self._e:stop()
+function Event:add(listener)
+  table.insert(self._listeners, listener)
+end
+
+function Event:remove(listener)
+  utils.array_remove(self._listeners, listener)
+  if #self._listeners == 0 then
+    self:destroy()
+  end
+end
+
+function Event:destroy()
+  log.line("watcher", "Event:destroy '%s'", self._path)
+
+  if self._fs_event then
+    local rc, _, name = self._fs_event:stop()
     if rc ~= 0 then
-      utils.notify.warn(
-        string.format("Could not stop the fs_event watcher for path %s : %s", self._opts.absolute_path, name)
-      )
+      utils.notify.warn(string.format("Could not stop the fs_event watcher for path %s : %s", self._path, name))
     end
-    self._e = nil
+    self._fs_event = nil
   end
-  for i, w in ipairs(M._watchers) do
-    if w == self then
-      table.remove(M._watchers, i)
-      break
-    end
+
+  Event._events[self._path] = nil
+end
+
+function Watcher:new(path, callback, data)
+  log.line("watcher", "Watcher:new '%s'", path)
+
+  local w = setmetatable(data, Watcher)
+
+  w._event = Event._events[path] or Event:new(path)
+  w._listener = nil
+  w._path = path
+  w._callback = callback
+
+  if not w._event then
+    return nil
+  end
+
+  w:start()
+
+  table.insert(Watcher._watchers, w)
+
+  return w
+end
+
+function Watcher:start()
+  self._listener = function()
```

Comment:
I wanted to call back to a member function like `Watcher:on_event` however it doesn't seem possible.